service/controller/clusterapi/v29/resource/clusterazs/create.go

```diff
@@ -48,30 +65,155 @@ func (r *Resource) EnsureCreated(ctx context.Context, obj interface{}) error {
+			}
 		}
 	}
 
 	{
+		// Parse TCCP network CIDR.
+		_, clusterCIDR, err := net.ParseCIDR(key.StatusClusterNetworkCIDR(cr))
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		// Split TCCP network between maximum number of AZs.
+		clusterAZSubnets, err := ipam.Split(*clusterCIDR, key.MaximumNumberOfAZsInCluster)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		// Convert collected availability zones into controllercontext types
+		// and allocate AZ level subnets when needed.
+		var ccAZs []controllercontext.ContextStatusTenantClusterAvailabilityZone
+		for az, subnets := range azs {
+			ccAZ := controllercontext.ContextStatusTenantClusterAvailabilityZone{
+				Name: az,
+			}
+
+			// Check if subnets of given availability zone already contain
+			// value?
+			if !subnets.areEmpty() {
+				ccAZ.PublicSubnet = subnets.public
+				ccAZ.PrivateSubnet = subnets.private
+
+				// Calculate the parent network from public subnet (always
+				// present for functional AZ).
+				parentNet := ipam.CalculateParent(subnets.public)
+
+				r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("availability zone %q already has subnet allocated: %q", az, parentNet.String()))
+
+				// Filter out already allocated AZ subnet from available AZ
+				// size networks.
+				clusterAZSubnets = ipam.Filter(clusterAZSubnets, func(n net.IPNet) bool {
+					return !reflect.DeepEqual(n, parentNet)
+				})
+			} else {
+				if len(clusterAZSubnets) > 0 {
+					// Pick first available AZ subnet and split it to public
+					// and private.
+					subnets, err := ipam.Split(clusterAZSubnets[0], 2)
+					if err != nil {
+						return microerror.Mask(err)
+					}
+
+					r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("availability zone %q doesn't have subnet allocation - allocated: %q", az, clusterAZSubnets[0].String()))
+
+					// Update available AZ subnets.
+					clusterAZSubnets = clusterAZSubnets[1:]
+
+					// Persist allocated & split subnets.
+					ccAZ.PublicSubnet = subnets[0]
+					ccAZ.PrivateSubnet = subnets[1]
+				} else {
+					return microerror.Maskf(invalidConfigError, "no more unallocated subnets left but there's this AZ still left: %q", az)
+				}
+			}
+
+			ccAZs = append(ccAZs, ccAZ)
+		}
+
 		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("setting cluster availability zones to controllercontext: %#v", azs))
 
-		cc.Status.TenantCluster.AvailabilityZones = azs
+		cc.Status.TenantCluster.AvailabilityZones = ccAZs
 	}
 
 	return nil
 }
+
+// fromEC2SubnetsToMap extracts availability zones and public / private subnet
+// CIDRs from given EC2 subnet slice and returns respectively structured map or
+// error on invalid data.
+func fromEC2SubnetsToMap(ss []*ec2.Subnet) (map[string]subnetPair, error) {
+	azMap := make(map[string]subnetPair)
+
+	for _, s := range ss {
+		if s == nil || s.AvailabilityZone == nil || s.CidrBlock == nil || s.Tags == nil {
```

Comment:
Why do we continue here instead of error? Is there a case where this would be fine?

Reply:
My expectations towards that pile of pointers and input coming from outside of this resource are low. I don't see problem with skipping invalid entries here when functionality depends on actually gathered information.

Approach to these is always a double edged sword: If one errors here, it could be that for reason or another the slice contains valid entries that are just invalid in this use case and that prevents further reconciliation while everything would be fine. Alternatively we might skip them and there's possibility that everything goes fine but something gets omitted because of skipping something fishy here.

That being said, I'm fine returning error here if that makes sense for you.

Reply:
My thinking would be that the data we manage internally should be considered complete. If information is missing the resource should be canceled early. Otherwise we should be able to expect perfect information at this point. Everything else is a bug that we need to fix and verify with unit tests.

Reply:
This returns an error now.